### ISSUE TITLE
Diffie Helmann information for each cipher suite no longer returned

### ISSUE THREAD

**Issue by BenjaminHae** [nabla-c0d3/sslyze#331]
In version 2.0.0 the `dh_info` key for ciphers that use Diffie-Hellmann key exchange has disappeared.
Are you planning to bring it back, or is it gone for good?

Example JSON from `1.4.3`:
```
    "tlsv1": {
                    "accepted_cipher_list": [
                        {
                            "dh_info": {
                                "A": "0x00ffffffff00000001000000000000000000000000fffffffffffffffffffffffc",
                                "B": "0x5ac635d8aa3a93e7b3ebbd55769886bc651d06b0cc53b0f63bce3c3e27d2604b",
                                "Cofactor": "1",
                                "Field_Type": "prime-field",
                                "Generator": "0x046b17d1f2e12c4247f8bce6e563a440f277037d812deb33a0f4a13945d898c2964fe342e2fe1a7f9b8ee7eb4a7c0f9e162bce33576b315ececbb6406837bf51f5",
                                "GeneratorType": "uncompressed",
                                "GroupSize": "256",
                                "Order": "0x00ffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551",
                                "Prime": "0x00ffffffff00000001000000000000000000000000ffffffffffffffffffffffff",
                                "Seed": "0xc49d360886e704936a6678e1139d26b7819f7e90",
                                "Type": "ECDH"
                            },
                            "is_anonymous": false,
                            "key_size": 256,
                            "openssl_name": "ECDHE-RSA-AES256-SHA",
                            "ssl_version": "TLSV1"
                        },
```

Example from `2.0.0`:

```
                        {
                            "is_anonymous": false,
                            "key_size": 256,
                            "openssl_name": "ECDHE-RSA-AES256-SHA",
                            "ssl_version": "TLSV1_1"
                        },
```

**Comment by nabla-c0d3** [comment]
I want to bring it back but the previous code was to difficult to maintain.

**Comment by faldridge** [comment]
I need this as well. Was the problem with both the C code and the Python usage of it?

I ask because I'm interested in submitting a PR and was wondering if simply restoring the low-level Python extensions, and then using them in the new Python SSL client code to be a valid approach.

**Comment by nabla-c0d3** [comment]
The problem was that the code that was reading the DH information using "private" APIs and data structures; hence it would only work on specific versions of OpenSSL, and it also required some heavy lifting to be done during the build process. 
When I had to add OpenSSL 1.1.1 (for TLS 1.3 support) in the new version of SSLyze, the DH code did not work with that version of OpenSSL. Then I decided to remove the DH code.

The right way to bring the functionality back would be to find an "official"/public API to retrieve the DH info in OpenSSL 1.1.1.

**Comment by faldridge** [comment]
Thanks for the update! I'll try to work on this as I have time.

**Comment by SQ-SEN** [comment]
A fix for this would be really appreciated. Currently I have to run all tests twice - with 1.4.1 and 2.0.1. A lot of setups have issues with DH param strength...

**Comment by ninijay** [comment]
Bump, because I would really appreciate this fix

**Comment by ninijay** [comment]
Also, the 1.1.1 manpages have some info about DH: 

* https://www.openssl.org/docs/man1.1.1/man3/DH_check_params.html
* https://www.openssl.org/docs/man1.1.1/man3/DH_check.html

I'm not sure what endpoint you'll need for sslyze, but maybe you can find something useful in [these search results](https://www.google.com/search?biw=1920&bih=1099&ei=tzdEXZaDJfDErgShjJLoCw&q=DH+%22man1.1.1%22+site%3Awww.openssl.org&oq=DH+%22man1.1.1%22+site%3Awww.openssl.org&gs_l=psy-ab.3...4312.5576..5757...0.0..0.104.412.4j1......0....1..gws-wiz.......0i71.iI8rZjpUEHo&ved=0ahUKEwjWurXZouTjAhVwoosKHSGGBL0Q4dUDCAo&uact=5).

**Comment by botastic** [comment]
I'd also really appreciate some more information on the key exchange and the authentication used.

**Comment by botastic** [comment]
Correct me if I'm wrong:
According to the [RFC](https://tools.ietf.org/html/rfc5246#section-7.4.2), the server **must** provide a certificate that is meeting the requierments of the negotiated cipher suite. Therefore the server **has to have** different certificates if multiple cipher suites with different key exchange algorithms are supported. SSLyze only returns one certificate chain.

If we'd check the certificate for every accepted cipher suite, then we can get the public key and information about the key exchange (and authentication) from the certificate and don't depend on OpenSSL.

This would of course come with the overhead of requesting a lot more certificate chains, but this also seems to be the "correct" implementation to me.

I think we would need to create a new plugin as the certificate and cipher suite plugins are currently indepent, but I just had a quick look.

**Comment by BenjaminHae** [comment]
@botastic  a server usually only supports one authentication algorithm- In most cases this is RSA, in some cases it is ECDSA.
Ciphers starting with `TLS_DHE_RSA_WITH_` use the DHE key exchange but RSA as authentication. So even if DHE is used for key exchange, the certificate will contain an RSA public key.
I don't remember ever seeing a server supporting ciphers matching `TLS_*_RSA_WITH_*` and `TLS_*_ECDH_WITH_*` at the same time. Only in that case 2 certificates are needed. 
In most cases only one certificate is necessary **and possible** as either ECDSA or RSA are supported by the server as authentication mechanism.

**Comment by botastic** [comment]
Unfortunately I don't, but it really would be interesting.

Nonetheless, wouldn't it still be better to check all certificates or supply an option for it? If different certificates are allowed for different cipher suites with the same authentication method, one of them could have a weak RSA (/ECDSA/DSS) key while the other ones have perfectly fine ones. Admittedly this is rather unlikely.

**Comment by BenjaminHae** [comment]
@botastic I just had someone look into it: If there are two certificates (ECDSA and RSA), only one of them is shown in the results (console or JSON output). Testssl however shows both certificates.

**Comment by r0oth3x49** [comment]
@nabla-c0d3 can you please guide me on why legacy ssl client's ServerKeyExchange bytes are different.

I have used openssl (1.0.2e i guess) compiled by testssl. the ServerKeyExchange bytes it shows are as follow:

set cipher to --> *ECDHE-RSA-AES256-GCM-SHA384* or *kECDHE*
target :- *microsoft.com*

```

<<< TLS 1.2 Handshake [length 019d], ServerKeyExchange
    0c 00 01 99 03 00 0e 91 04 05 da 1d 8d 50 98 1d
    3c 0b 8c 27 f9 34 75 e2 ff a0 07 6e 48 c9 ba bf
    b5 18 17 d7 32 8c 91 95 c3 32 ea 45 68 4b 0a d8
    c6 e1 80 e0 6b 97 37 66 9c d0 5e bc 9a bb b8 ba
    2f f9 ab 3e 14 4f 27 b9 63 70 4a 10 f2 1e 6b b7
    c6 02 84 d5 cc f9 15 8d e1 10 d1 1f 34 49 52 ed
    d4 c0 3c 70 e7 ca 11 a4 b7 4b 86 b0 ab 6a c8 f5
    8e 31 05 0b 8d 04 9c 4f 94 91 3d 12 89 2e c0 84
    42 aa ea e8 f5 02 00 0a e4 76 0c d8 ee 22 a6 dd
    60 1b 15 29 cd dd 4b a9 c9 06 01 01 00 49 99 4c
    13 ec 7b 04 b2 cd ce 21 c0 47 3f f8 88 bb 1b 3e
    8f e2 be 38 e8 30 4d 72 bd bf 42 e6 51 c0 d9 7e
    81 26 f9 50 03 5f 58 4a aa 40 b2 4b a7 4e b7 30
    b0 66 7b 5b 1d 7d c5 1b da 92 8a 4c 88 f0 b6 ec
    23 3d 7d f3 95 7c f7 26 f6 50 cc 41 39 a1 27 71
    5d 2c 91 f2 0f 9b 0d 4a 8e c7 f5 0b 13 71 90 6d
    3c 18 93 8f 3f 54 f3 2b be df 99 97 9e d7 03 59
    7d 81 ef 19 12 f1 2d 49 f4 c5 ae dd ee ef 7c b7
    30 5d 70 a8 9a 73 6e f2 3b 42 41 d6 37 0c 66 cd
    fd dc 8a ef 95 14 ce a9 ed 02 13 f5 ff 4b 0a a5
    59 44 69 e8 94 0d 41 46 17 6c c4 ab 03 8f 24 a3
    7d 9e ac 38 91 ad a8 78 e4 cf b3 c0 f5 51 ab a6
    6d 75 78 f0 e6 5c 4c 41 55 0e 0d a7 96 9b f1 ed
    f7 10 12 ef a0 53 46 d7 cc 1a 8b 8c 41 17 0a 7a
    62 c9 25 e3 51 8f 17 85 54 91 1d bc 5f 85 ba 36
    5e ef 22 4c d4 20 c6 65 0c 9a c0 ec 77
<<< ??? [length 0005]
``` 

While you legacy client for same cipher using SSlyze `2.1.4` the ServerKeyExchange bytes are different. 

```
03001741046218909672219b1b34dfe2c2601e678a519d00cbbfe69ec482772b3bafe03a1aece7cc146b447ff6634900f9785b7cd63a9bf836660e65cd4e5feca4e99d159b06010100c4714a88efa50c342c967e4e55ad6848a4d5c421f514eaa58027a2642327de7c99ca0b9f870e09b02c8e2416edbffb81fc33aecbc91a95e745a0933d1f3828eecbafa6ddce2d5a68de9622816f465522fc90a27432edcfea79c25140b9b436d97e561d8294e93640f54c7b40df93340b72473722e2a04eb18cf52cf88d0b5b962d6523934d065c474c08d946ca97d255e7ef742f1b49e3ade68b5138fd80d6b46919b9e6cbc7c90c861b9ba7d803698327bc20d4836da08182f762ac8ab2634633a855698f6eb12d6bf2a4e42e38288dede8657477b3c183f1897b724379cfd6a7c432db40d33cf05cee804e3ca7768bf9996e242eefbfeca302bde34e933c38
```

my perticular concern is to find ECDH within KeyExchange records.

which in both cases are differnt.

Using SSLyze Legacy SslClient with TLS1.2.
 - the bytes are --> *03 00 17* which is: prime256v1.

while according to testssl it is *03 00 0e* which is: sect571r1

so can you please take a look what's the issue or 'm i getting it wrong?

**Comment by r0oth3x49** [comment]
we can still calculate DH information from Handshake bytes

**Comment by nabla-c0d3** [comment]
@FWinterborn opened some really cool pull requests for this:

* https://github.com/nabla-c0d3/nassl/pull/48
* https://github.com/nabla-c0d3/sslyze/pull/394

I ll try to get them released soon.

**Comment by fabian-hk** [comment]
I am really interested in the changes @FWinterborn made because I need them for my Bachelor Thesis project. Is there a timetable when this is going to be released? And is there something I can help you with?

**Comment by nabla-c0d3** [comment]
Fixed in v3.0.0.
